File: packages/sdk-ruby/lib/melaya/projects.rb
# frozen_string_literal: true

module Melaya
  # Projects API — create and list agent projects.
  #
  # Projects are the top-level namespace for pipelines, connectors, and team
  # membership in the Melaya platform.
  #
  # Maps to /api/v1/private/projects/*.
  #
  # @example
  #   projects = melaya.projects.list
  #   project  = melaya.projects.create(name: "my-project", description: "...")
  #   melaya.projects.rename(old_name: "my-project", new_name: "renamed-project")
  class ProjectsAPI
    def initialize(http)
      @http = http
    end

    # GET /api/v1/private/projects
    # List all projects the authenticated user can access
    # (owned projects + projects they are a member of).
    def list
      @http.get("/api/v1/private/projects")
    end

    # POST /api/v1/private/projects
    # Create a new agent project.
    # @param name [String]
    # @param description [String, nil]
    def create(name:, description: nil)
      body = compact("name" => name, "description" => description)
      @http.post("/api/v1/private/projects", body)
    end

    # PATCH /api/v1/private/projects/rename
    # Rename a project (body: oldName, newName).
    # @param old_name [String]
    # @param new_name [String]
    def rename(old_name:, new_name:)
      @http.patch("/api/v1/private/projects/rename",
        "oldName" => old_name,
        "newName" => new_name)
    end

    # GET /api/v1/private/projects/runner
    # Get projects list (runner-facing, authed).
    def runner_projects
      @http.get("/api/v1/private/projects/runner")
    end

    private

    def compact(hash)
      hash.reject { |_, v| v.nil? }
    end
  end
end

File: packages/sdk/src/projects.ts
/**
 * Projects API — create and list agent projects.
 *
 * Maps to `/api/v1/private/projects`. Projects are the top-level namespace
 * for pipelines, connectors, and team membership.
 */
import type { HttpClient } from "./client.js";
import type { Project, ProjectCreate } from "./platform-types.js";

export class ProjectsAPI {
  constructor(private readonly http: HttpClient) {}

  /**
   * List all projects the authenticated user can access
   * (owned projects + projects they are a member of).
   */
  async list(): Promise<Project[]> {
    return this.http.get<Project[]>("/api/v1/private/projects");
  }

  /** Create a new agent project. */
  async create(body: ProjectCreate): Promise<Project> {
    return this.http.post<Project>("/api/v1/private/projects", body);
  }

  /** Rename a project. */
  async rename(oldName: string, newName: string): Promise<{ ok: boolean }> {
    return this.http.patch<{ ok: boolean }>("/api/v1/private/projects/rename", {
      oldName,
      newName,
    });
  }

  async runnerProjects(): Promise<Project[]> {
    return this.http.get<Project[]>("/api/v1/private/projects/runner");
  }
}
